### PR TITLE
New module IMPCAP & ommongodb improvement

### PR DESCRIPTION
**Pull request by KGuillemot**
#### Add new modules : 
 - **impcap** : Capture packets from network interfaces and play with them in rsyslog as JSON objects
 - **mmcapture** : Extract payloads of ipcap captured packets and store them on disk

#### Improve ommongodb : 
 - Let ommongodb module works event if mongo-c-driver is not compiled with SSL support.
 - Add parsing of "date" or "time" field as rfc3339 to insert as ISODate in MongoDB

### CODE REVIEW

**Comment by rgerhards** [comment]
Hi, thx for the PR, much appreciated. I will have a deeper look soon. In the meantime, could you consider squashing the commits? Best would be one commit per new module plus the enhancement, but I do not know if that is possible.

We also need some docs (see https://github.com/rsyslog/rsyslog-doc).

**Comment by KGuillemot** [comment]
Hi, you're wellcome, that improvement is really needed for our use of Rsyslog, and I think it will be very helpful for everybody :-)

I'll try to squashing the commits, but I don't know for now how to do.

I'll do that after solving the "codecov/project" test failure, do you have any idea on what goes wrong ?

Thanks

**Comment by KGuillemot** [comment]
I really tried fixing the issue concerning "codecov/project" but I don't quietly understand the changes showed by the website, for example : https://codecov.io/gh/rsyslog/rsyslog/compare/ddacd3042131dbdec4abbfcb02a214f166b10e6a...0ea25525192ef15e5e03bd093e27499e046bb9ce/src/plugins/imjournal/imjournal.c

I never modified this file, and when I check on my repository, the last commit was on 26 Oct 2018.

Could you please help me ?

**Comment by rgerhards** [comment]
> I'll try to squashing the commits, but I don't know for now how to do.

There are various ways - I usually use `git rebase`. You may want to have a look e.g. at https://gist.github.com/aortbals/2aeb557bf127dd7ae88ea63da93479fc

For new modules, you can also start with a fresh copy of master branch, and just copy the new modules and changes over. After that, make fixes not be `git commit` but `git commit --amend` so that they are put into the same commit.

Ideally, we have different commits for different branches, so that it is easy to cherry-pick (e.g. for distro maintainers). It makes it also much easier to do troubleshooting via the git history. But it may be too late now ;-)

If all fails, I can squash myself and write back to your branch. Then, you could pick up the result from there. But then it would be just a single big commit, as I don't see the fine details.

Let me know which path you want to go forward.

**Comment by rgerhards** [comment]
I currently see no CodeCov result as the PR is running again. If it is a drop in Coverage, that's probably OK: we have enabled compilation of a contributed module in order to get a enhancement PR fully checked. As of now, that module has no tests, so Coverage drops. As soon as CI is done again, I'll check and let you know if somethings needs your attention.

**Comment by rgerhards** [comment]
... and at that instant Travis finished: it needs libpcap. Can you tell me what to install? I also need to add that to the containers used in buildbot-based CI.

**Comment by KGuillemot** [comment]
Thank you very much for your precious help, I really appreciate.

Yes I just added --enable-impcap and --enable-mmcapture in tests because I thank that codecov failed because some tests were missing.

On linux systems, the needed package is "libpcap-dev", on BSD systems the needed package is "libpcap".

**Comment by rgerhards** [comment]
I really appreciate your help and great collaboration!

> On linux systems, the needed package is "libpcap-dev"

For Travis, you can add this dependency yourself in `./test/travis/install.sh`. For the containers, I need to do that, But to avoid more confusion in other PRs, let's first get Travis going.

**Comment by rgerhards** [comment]
> The last failed build seems to be caused by a bug in Clang (v5.0.0) with __packed__.
> Please see the following post : https://stackoverflow.com/questions/48564212/clang-compaining-about-attribute-packed-even-though-the-struct-needs-to-b

Thx for the pointer. So it looks like we need to work-around it, which thankfully is easy to do. We can use a `#pragma` to disable the warning if clang 5 is detected. Something similar is done e.g. here: https://wordpress.org/plugins/wp-vr-view/ (except it does not check the version and push/pop the warning stack). If you like, I can provide concrete code.

**Comment by KGuillemot** [comment]
Ok, I'll check that.

I think that your link is a mistake ^^

**Comment by rgerhards** [comment]
> I think that your link is a mistake ^^

Indeed ;-) I just fixed it, correct one was https://github.com/rsyslog/rsyslog/blob/master/plugins/imfile/imfile.c#L1182

**Comment by rgerhards** [comment]
side-note: I would really like to have tests for ommongodb inside CI. Could we work together on getting them done - maybe on a different PR?

**Comment by KGuillemot** [comment]
> side-note: I would really like to have tests for ommongodb inside CI. Could we work together on getting them done - maybe on a different PR?

Yes of course ! 

Concerning the 2 Travis failures, I defined #pragma for the \_\_packed\_\_ attributes but there is still some errors I can't solve because of the implementation of the algorithm.
For example : 
```arp_parser.c:64:29: error: cast from 'const uchar *' (aka
      'const unsigned char *') to 'arp_header_t *' (aka 'struct arp_header_s *')
      increases required alignment from 1 to 2 [-Werror,-Wcast-align]
        arp_header_t *arp_header = (arp_header_t *)packet;
                                   ^~~~~~~~~~~~~~~~~~~~~~
```
I can't solve that warning because we **need** to do that...

An other error is : 
```impcap.c:481:6: error: no previous prototype for function
      'init_eth_proto_handlers' [-Werror,-Wmissing-prototypes]
void init_eth_proto_handlers() {
```
which is not correct because we do define the prototype upper in the file.

I don't really know how to solve this errors.
Maybe with #pragma ?

**Comment by rgerhards** [comment]
Let's look at this one first:

> An other error is :
> 
> ```
>       'init_eth_proto_handlers' [-Werror,-Wmissing-prototypes]
> void init_eth_proto_handlers() {
> ```

The compiler is actually right. It also emits this message:
```
impcap.c:481:6: error: no previous prototype for function
      'init_eth_proto_handlers' [-Werror,-Wmissing-prototypes]
void init_eth_proto_handlers() {
     ^
impcap.c:77:6: note: this declaration is not a prototype; add 'void' to make it
      a prototype for a zero-parameter function
```

This is indeed not a prototype but an old-style definition which says nothing about the parameter types. I assume you want to have no parameters. The you need to say:
```
void init_eth_proto_handlers(void) {
```
Now it tells that this function accepts no (void) arguments.

I need to look into the alignment, but it is a proper error, I think. ARM will probably error out. I guess the rest of the CI just runs because it does not yet compile the modules. I'll see if I can add this without breaking too many other builds.

On the alignment I think I know the solution, but I need to check. Again, we should not take that lightly, IIRC it's a real problem on some archs.

**Comment by rgerhards** [comment]
note: I am working on ARM-based CI; may fail with missing packages

**Comment by rgerhards** [comment]
as I thought: on the rapsbian machine the alignment is a hard error - see CI (the arm is still invalid, package missing)

**Comment by rgerhards** [comment]
see https://github.com/rsyslog/rsyslog/pull/2509 for a possible solution for the alignment issue (tl;dr: use `union`).

File: plugins/impcap/impcap.c [hunk 1]
```diff
@@ -599,12 +599,13 @@ void packet_parse(uchar *arg, const struct pcap_pkthdr *pkthdr, const uchar *pac
...
 void *startCaptureThread(void *instanceConf) {
 	int id = 0;
-	instanceConf_t * inst = (instanceConf_t * )
-	instanceConf;
+	instanceConf_t * inst = (instanceConf_t * )instanceConf;
 	while (glbl.GetGlobalInputTermState() == 0) {
 		pcap_dispatch(inst->device, inst->pktBatchCnt, packet_parse, (uchar * ) & id);
 	}
 	pthread_exit(0);
+	/* the function must return something - NULL will do */
```
Comment (rgerhards) [inline plugins/impcap/impcap.c]: Actually, the function never reaches that point. Pragma `noreturn` is a better way to handle that situation. We have a mactro `ATTR_NORETURN` that applies it depending on if the compiler supports it or not. I'll suggest it above.

File: plugins/impcap/impcap.c [hunk 2]
```diff
@@ -599,12 +599,13 @@ void packet_parse(uchar *arg, const struct pcap_pkthdr *pkthdr, const uchar *pac
 */
 void *startCaptureThread(void *instanceConf) {
```
Comment (rgerhards) [inline plugins/impcap/impcap.c]:
```suggested
void *startCaptureThread(void *instanceConf) ATTR_NORETURN {
```

File: plugins/mmcapture/mmcapture.c [hunk 3]
```diff
@@ -60,7 +60,7 @@ DEF_OMOD_STATIC_DATA
 #define IMPCAP_DATA     "!data"
 #define DEFAULT_LOG_DIR "/var/log/rsyslog/"
 
-static char *proto_list[] = {
+static const char proto_list[][] = {
```
Comment (rgerhards) [inline plugins/mmcapture/mmcapture.c]: just FYI: `const` was missing, `[]` is not necessary (but ok).

**Comment by KGuillemot** [comment]
Thank you for your help, I did the ATTR_NORETURN modification.

**Comment by rgerhards** [comment]
> Thank you for your help, I did the ATTR_NORETURN modification.

Thanks for your persistence! I have also seen that the compilers do not like the new 2d array definition. I'd simply go back to `char *` but with `const`. I'll try to keep an eye to help with nits when I see them.

Note: one of the arm CI envs is still missing libpcap, so it will err out on that. Hope to fix it soon.

**Comment by KGuillemot** [comment]
> > Thank you for your help, I did the ATTR_NORETURN modification.
> 
> Thanks for your persistence! I have also seen that the compilers do not like the new 2d array definition. I'd simply go back to `char *` but with `const`. I'll try to keep an eye to help with nits when I see them.
> 
> Note: one of the arm CI envs is still missing libpcap, so it will err out on that. Hope to fix it soon.

Thank you very much !

I had a compilation error with char * : 
```
mmcapture.c:64:2: error: initializing 'char *' with an expression of type 'const char [5]' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        "http",
        ^~~~~~
mmcapture.c:65:2: error: initializing 'char *' with an expression of type 'const char [4]' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        "ftp",
        ^~~~~
mmcapture.c:66:2: error: initializing 'char *' with an expression of type 'const char [4]' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        "smb"
        ^~~~~
```

**Comment by rgerhards** [comment]
> I had a compilation error with char * :

yup - these are actually constant strings (you cannot modify them). So it needs to be `const char *`. The `const` was missing.

**Comment by KGuillemot** [comment]
I don't really understand were is the error in the failed build "debian9 rsyslog" : 
https://build.rsyslog.com/#/builders/64/builds/1754

Could you please check it out ?

File: plugins/mmcapture/smb_sessions.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ *       -or-
+ *       see COPYING.ASL20 in the source distribution
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
Comment (rgerhards) [inline plugins/mmcapture/smb_sessions.h]:
```suggested
#include "config.h"
```
This include must be the first include in all source files. This is an autoconf requirement. I guess this also triggers the `strndup()` error.

**Comment by KGuillemot** [comment]
I just fixed `strndup` error by adding `#include "config.h` at the top of source files as you recommended me.

This seems to work well, but I still have an error concerning the \_\_packed\_\_ attribute : 
```
eth_parser.c:40:1: error: packed attribute is unnecessary for 'eth_header_s' [-Werror=packed]
 };
 ^
```

I can't say if this attribute is really *unnecessary* but I don't really want to remove it, we never know..

Do you think the best solution would be to remove the `#if defined(__clang__)` to set the #pragma on gcc **and** clang ?

**Comment by rgerhards** [comment]
> Do you think the best solution would be to remove the `#if defined(__clang__)` to set the #pragma on gcc **and** clang ?

sounds like a good plan to me - I don't have a better suggestion right now (maybe we just both don't know the real cure, but we can always change if we learn it!).

**Comment by rgerhards** [comment]
FYI: I have now updated the so-far missing arm machine (raspbian).

**Comment by KGuillemot** [comment]
Thank you.

I still have a compilation error concerning packed, I'll have to add `-Werror=packed` ...

Error : 
```
eth_parser.c:49:1: error: packed attribute is unnecessary for 'vlan_header_s' [-Werror=packed]
 };
```
I wait your answer before doing the modification.

**Comment by rgerhards** [comment]
I haven't tried to find the instance with clang, but I see it with gcc8 as well. Maybe you need to disable it for both?

**Comment by rgerhards** [comment]
In case you haven't seen, the static analyzer also shows problems.

**Comment by KGuillemot** [comment]
There seems to be a problem in the **continuous-integration/travis-ci/pr** : https://travis-ci.org/rsyslog/rsyslog/jobs/492293695

I let you check but I don't think it is related to my commits..

Thanks

**Comment by rgerhards** [comment]
`es-bulk-retry` this is known to cause false positives. If only it is left, let me handle it. But I see there are compile errors right now.

File: plugins/impcap/impcap.c [hunk 5]
```diff
@@ -576,11 +576,11 @@ void packet_parse(uchar *arg, const struct pcap_pkthdr *pkthdr, const uchar *pac
 	struct tm timestruct;
 	time_t rawtime = time(NULL);
 	gmtime_r(&rawtime, &timestruct); // Get the current UTC date
-	strftime(strtime, 30, "%Y-%m-%dT%H:%M:%S%Z", timestruct); // Format the time as UTC str
```
Comment (rgerhards) [inline plugins/impcap/impcap.c]: just a reminder: `strftime()` is not  safe to use as it depends on the "TZ" environment variable.

Comment (KGuillemot) [reply plugins/impcap/impcap.c]: Ok thank you.
What can I use safely instead ? 
I have to format the date as UTC to be compliant with MongoDB ISODate, that I use to send my impcap logs to.

Comment (rgerhards) [reply plugins/impcap/impcap.c]: Unfortunately, there is no simple solution. IMHO linux timezone and time manipulation API is severely broken. I just wanted to make you aware of the fact. In our code, we do all the conversions ourselfs, but that limits us in regard to support for different TZ. :-(

**Comment by rgerhards** [comment]
The codecov drop means we need tests for the new features. Please check Codacy - not sure if these are valid concerns, if not, we need to see how to handle them at the backend layer.

**Comment by rgerhards** [comment]
@KGuillemot The remaining codacy failures look valid to me. Otherwise I think we made a big step forward :-)

**Comment by KGuillemot** [comment]
> @KGuillemot The remaining codacy failures look valid to me. Otherwise I think we made a big step forward :-)

Yes it seems valid to me too, I didn't quite understand what was wrong ^^
That's great ! 

Also, I'm not comfortable with the codecov interface, I just don't understand it ... I need to add some tests related to impcap ?

**Comment by rgerhards** [comment]
> Also, I'm not comfortable with the codecov interface, I just don't understand it ... I need to add some tests related to impcap ?

Give me a bit, I'll write it up.

**Comment by rgerhards** [comment]
@KGuillemot hope this helps: https://rainer.gerhards.net/2019/02/understanding-codecov-reports.html

Questions on test plumbing, please let me know, happy to discuss!

**Comment by KGuillemot** [comment]
> > Also, I'm not comfortable with the codecov interface, I just don't understand it ... I need to add some tests related to impcap ?
> 
> Give me a bit, I'll write it up.

I am very grateful to you, what do you need exactly to wrote some tests ?

> hope this helps: https://rainer.gerhards.net/2019/02/understanding-codecov-reports.html
> Questions on test plumbing, please let me know, happy to discuss!

I'll check that :-)

I also add an improvement in OmMongoDB, see my last commit, to support date-rfc3339 format, for example used by : `set $!time = format_time($!timestamp_app, "date-rfc3339");`
I use it like that to send my PacketFilter logs into Mongodb (epoch format).

**Comment by rgerhards** [comment]
@KGuillemot the current failures are false positives. I am right now looking into it, for some reason some clickhouse tests suddenly do no longer work. But it would be good if you could fix the Codacy issues.

**Comment by rgerhards** [comment]
I could solve the CI flakes, rest is valid.

**Comment by rgerhards** [comment]
@KGuillemot do you expect to work on this PR this week? 8.1903.0 release window closes friday, Just FYI.

**Comment by KGuillemot** [comment]
Hello,
I'm really sorry I was just not available.
>Give me a bit, I'll write it up.

What do you need, and what can I do ?

**Comment by rgerhards** [comment]
The write-up is alresdy done see above: https://github.com/rsyslog/rsyslog/pull/3432#issuecomment-464660297

**Comment by KGuillemot** [comment]
Thank you, so what can I do now ?

**Comment by rgerhards** [comment]
Pls fix the errors shown in CI.

Sent from phone, thus brief.

Kevin Guillemot <notifications@github.com> schrieb am Mo., 11. März 2019,
13:38:

**Comment by rgerhards** [comment]
Better said: check the conversation. Made a rebase AVD other things are to
be done.

Sent from phone, thus brief.

Rainer Gerhards <rgerhards@hq.adiscon.com> schrieb am Mo., 11. März 2019,
13:53:

**Comment by rgerhards** [comment]
@KGuillemot do you intend to finish this PR? It's been a lot of work up to here and I think it would be a shame to just close it unmerged...

**Comment by rgerhards** [comment]
Please fix these issues: https://app.codacy.com/app/rsyslog/rsyslog/pullRequest?prid=2913093

**Comment by KGuillemot** [comment]
> Please fix these issues: https://app.codacy.com/app/rsyslog/rsyslog/pullRequest?prid=2913093

It seems that these errors are not really errors. 

It's because I define my pointers as NULL before calling, for example, `json_object_new_object()`. It's a security.

Do you want me to do this : 
```
struct json_object *json = json_object_new_object();
```
instead of this ?
```
struct json_object *json = NULL;
json = json_object_new_object();
```

**Comment by rgerhards** [comment]
> It seems that these errors are not really errors.

False positives are often the case with analyzers. We have the policy that these false positives must be worked around in order to avoid continuous confusion and CI failure. It's explained here in the context of static analysis: https://rainer.gerhards.net/2018/06/why-static-code-analysis.html

> It's because I define my pointers as NULL before calling, for example, `json_object_new_object()`. It's a security.
> 
> Do you want me to do this :
> 
> ```
> struct json_object *json = json_object_new_object();
> ```
> instead of this ?
> 
> ```
> struct json_object *json = NULL;
> json = json_object_new_object();
> ```
definitely, that a) solves the issue and b) performs better.

**Comment by rgerhards** [comment]
as not only a side-note, we really need some dynamic tests for the modules. If there is no way around, it's acceptable to go without them for contributed modules, but it leaves some doubt. I'll check to see if I can contribute at least a handful of extremely basic one (aka "not really sufficient but at least a bit"). I am glad to help with better tests.

**Comment by rgerhards** [comment]
@KGuillemot you also need to move the new modules to the ./contrib tree as the rsyslog core team cannot maintain them

**Comment by rgerhards** [comment]
@KGuillemot I have contruibuted some very basic tests, see https://github.com/rsyslog/rsyslog/pull/3432/commits/25148ddebe30092be61d17841ae6eececb029b2a

It also looks like they detect some issues.

**Comment by KGuillemot** [comment]
> @KGuillemot I have contruibuted some very basic tests, see [25148dd](https://github.com/rsyslog/rsyslog/commit/25148ddebe30092be61d17841ae6eececb029b2a)
> 
> It also looks like they detect some issues.

Thank you very much, i'll check that.

**Comment by KGuillemot** [comment]
> Note: include to config.h is missing in that file. Must always be the first include in autoconf projects.

Yes that file is included in the parser.h file. The problem is there is 2 files called like that : 
 - rsyslog/runtime/parser.h
 - contrib/impcap/parser.h

The thing that I don't understand is that worked perfectly before I moved the directoriy to contrib insead of plugins.

**Comment by rgerhards** [comment]
@KGuillemot I fixed the problem with config.h, and now it seems to build. It's a hard requirement that this be the first include, see autoconf manual. I guess depending on platform they also setup some compiler options.

**Comment by rgerhards** [comment]
no - it doesn't build everywhere.. on CodeCov it now does. So let's see...

**Comment by KGuillemot** [comment]
I think it's an include issue but I don't really know how to fix it. Maybe in impcap Makefile ?

**Comment by rgerhards** [comment]
I am currently looking at it, have pulled the container and your code. I ack it is strange.

**Comment by rgerhards** [comment]
@KGuillemot stand by for the moment. I change CI defs to gain some more insight. Even in the container it builds.

**Comment by rgerhards** [comment]
oh wait - it's distcheck... I guess the include is not put into the tarball.

**Comment by rgerhards** [comment]
yup "make distcheck" reproduces the problem

**Comment by rgerhards** [comment]
@KGuillemot I think my commits sorted out the build issues. Please now have a look at the failing tests.

**Comment by KGuillemot** [comment]
I'll remove the mmdarwin module because it has not been fully tested and is not finished as we want to.
We want to improve-it before publishing it.
Do you mind ?

**Comment by KGuillemot** [comment]
I also had issue with my ommongodb improvement concerning the time : 
If the timezone is not UTC, the module fails and does not send to mongo...
Do you have any idea of what should I do ?  There is not so much logs even in debug mode, but I know it's my modification which causes the issue.

**Comment by rgerhards** [comment]
> I also had issue with my ommongodb improvement concerning the time :
> If the timezone is not UTC, the module fails and does not send to mongo...
> Do you have any idea of what should I do ? 

unforutnately not

> There is not so much logs even in debug mode, but I know it's my modification which causes the issue.

Regarding this, I'd say just add additonal debug info.

**Comment by rgerhards** [comment]
> I'll remove the mmdarwin module because it has not been fully tested and is not finished as we want to.
> We want to improve-it before publishing it.
> I'll rename the pull-request.

I guess I don't have the right context? This PR is not about mmdarwin or is it? I see https://github.com/rsyslog/rsyslog/pull/3556. Do you mean this one?

**Comment by KGuillemot** [comment]
> > I'll remove the mmdarwin module because it has not been fully tested and is not finished as we want to.
> > We want to improve-it before publishing it.
> > I'll rename the pull-request.
> 
> I guess I don't have the right context? This PR is not about mmdarwin or is it? I see #3556. Do you mean this one?

Sorry, I mean mmcapture module, not mmdarwin. I just modified my post.

**Comment by rgerhards** [comment]
> Sorry, I mean mmcapture module, not mmdarwin. I just modified my post.

ah, now I see. So you want to remove mmcapture from this PR here, but keep impcap and ommongodb improvements. Is this right? Feel free to do so. It would be useful if you keep any tests we have already developed.

Also, I would appreciate if you could squash and re-write the commit message to reflect what the PR is finally about. I admit I get a bit lost here - especially when it comes at what I need to add to the ChangeLog once done ;-)

**Comment by KGuillemot** [comment]
> > Sorry, I mean mmcapture module, not mmdarwin. I just modified my post.
> 
> ah, now I see. So you want to remove mmcapture from this PR here, but keep impcap and ommongodb improvements. Is this right? Feel free to do so. It would be useful if you keep any tests we have already developed.
> 
> Also, I would appreciate if you could squash and re-write the commit message to reflect what the PR is finally about. I admit I get a bit lost here - especially when it comes at what I need to add to the ChangeLog once done ;-)

Yes I understand.
Concerning the mmcapture module, i'll keep the tests and everything in our repository, in mmcapture branch to continue it and make a new pull-request later.
Concerning the ommongodb feature, did you see my post concerning issue with the timezone ?

For the squash, I never did that and I don't really want to break the pull-request by trying.
Could you please squash for me please ? I mean when all the build issues will be solved.

**Comment by rgerhards** [comment]
> Concerning the mmcapture module, i'll keep the tests and everything in our repository, in mmcapture branch to continue it and make a new pull-request later.
> Concerning the ommongodb feature, did you see my post concerning issue with the timezone ?

https://github.com/rsyslog/rsyslog/pull/3432#issuecomment-478488342

> For the squash, I never did that and I don't really want to break the pull-request by trying.
> Could you please squash for me please ? 

I can do this as part of the merge

> I mean when all the build issues will be solved.

Right now they are not, just to make sure we are on the same side.

**Comment by KGuillemot** [comment]
> > Concerning the mmcapture module, i'll keep the tests and everything in our repository, in mmcapture branch to continue it and make a new pull-request later.
> > Concerning the ommongodb feature, did you see my post concerning issue with the timezone ?
> 
> [#3432 (comment)](https://github.com/rsyslog/rsyslog/pull/3432#issuecomment-478488342)
> 
> > For the squash, I never did that and I don't really want to break the pull-request by trying.
> > Could you please squash for me please ?
> 
> I can do this as part of the merge
> 
> > I mean when all the build issues will be solved.
> 
> Right now they are not, just to make sure we are on the same side.

Yes we are, I'm currently working on.

**Comment by KGuillemot** [comment]
For the build **docker-centos7**, the test *es-basic-vg.sh* does not pass. 
And for the build **debian9**, the test *rcvr_fail_restore.sh* does not pass.
Is it normal ?

**Comment by rgerhards** [comment]
> For the build **docker-centos7**, the test _es-basic-vg.sh_ does not pass.
> And for the build **debian9**, the test _rcvr_fail_restore.sh_ does not pass.
> Is it normal ?

well "normal" ... of course flakes should not happen, but they seem to plague almost all CI systems. We try to find the issues, but so far did not succeed. Keep in mind that I usually babysit PRs and restart failures which seem to be flakes. The more busy CI gets, the more likely they are. This points towards races,

**Comment by KGuillemot** [comment]
I have an issue while trying to build on FreeBSD. 
I have the following error at the beginning of ./autogen.sh : 
```
Makefile.am:538: error: unterminated conditionals: ENABLE_DEFAULT_TESTS_TRUE
Makefile.am:17: warning: '%'-style pattern rules are a GNU make extension
automake: error: undefined condition 'TRUE' for 'RECURSIVE_TARGETS'
automake: RECURSIVE_TARGETS: 
automake:   {
automake:     ENABLE_DEFAULT_TESTS => {
automake:       type: +=
automake:       where: /usr/local/share/automake-1.16/am/texinfos.am:
automake:       comment: 
automake:       value: dvi-recursive html-recursive info-recursive pdf-recursive ps-recursive \
automake:   install-dvi-recursive \
automake:   install-html-recursive \
automake:   install-info-recursive \
automake:   install-pdf-recursive \
automake:   install-ps-recursive all-recursive check-recursive installcheck-recursive
automake:       owner: Automake
automake:     }
automake:   }
automake: 
automake: Please contact <bug-automake@gnu.org>.
 at /usr/local/share/automake-1.16/Automake/Channels.pm line 662.
	Automake::Channels::msg("automake", "", "undefined condition 'TRUE' for 'RECURSIVE_TARGETS'\x{a}RECURSIVE_"...) called at /usr/local/share/automake-1.16/Automake/ChannelDefs.pm line 212
	Automake::ChannelDefs::prog_error("undefined condition 'TRUE' for 'RECURSIVE_TARGETS'\x{a}RECURSIVE_"...) called at /usr/local/share/automake-1.16/Automake/Item.pm line 95
	Automake::Item::rdef(Automake::Variable=HASH(0x801ef92e8), Automake::Condition=HASH(0x8015edaf8)) called at /usr/local/bin/automake line 3972
	Automake::handle_subdirs() called at /usr/local/bin/automake line 7953
	Automake::generate_makefile("Makefile.am", "Makefile.in") called at /usr/local/bin/automake line 8231
	Automake::handle_makefile("Makefile.in") called at /usr/local/bin/automake line 8243
	Automake::handle_makefiles_serial() called at /usr/local/bin/automake line 8375
autoreconf-2.69: automake failed with exit status: 255
```
This error means something for you ?

**Comment by rgerhards** [comment]
any chance we can get some more tests?

**Comment by rgerhards** [comment]
@KGuillemot FYI: merge window for 8.1904.0 closes friday.

**Comment by KGuillemot** [comment]
It will be difficult for me to add some tests today.
I'll try to make some this week-end.
Do you have any test example for me please ?

**Comment by rgerhards** [comment]
@KGuillemot if you can tell me 
* what I need to install to have mongodb server installed
* how to start it up
* how to get data from it
* how to stop it
I could help with a based test for testing the modules functionality.

**Comment by rgerhards** [comment]
> Do you have any test example for me please ?

besides the MonogDB issue, tests reside in `./tests`. I can see that I add some very basic ones like https://github.com/rsyslog/rsyslog/blob/master/tests/improg_errmsg_no_params.sh

**Comment by rgerhards** [comment]
@KGuillemot oh, I already added very basic tests. check `./tests/impcap_errmsg_no_params*sh`

**Comment by KGuillemot** [comment]
> @KGuillemot oh, I already added very basic tests. check `./tests/impcap_errmsg_no_params*sh`

Yes I saw this tests, thank you for that !
For MongoDB you just have to install the package, I can give you a configuration file with tls verification to test.
For impcap, the 2 tests you already added are enought ?

**Comment by rgerhards** [comment]
> For MongoDB you just have to install the package, I can give you a configuration file with tls verification to test.

Please let me know the package names for at least Ubuntu, CenotOS 7 and Fedora. For the containers, the server must be started and stopped during the test. So I need to have a way to do that.  The config is a good start, but it's just a start.

> For impcap, the 2 tests you already added are enought ?

Not really, better than nothing but they don't really test anything of interest. I think code statment coverage is around 15% ;-) I don't know if it is possible to actually test more, but you as module developer should know.

**Comment by rgerhards** [comment]
@KGuillemot any interest in finishing this PR?

**Comment by rgerhards** [comment]
this looks abandanoned. Would be great if someone would pick it up.

**Comment by frikilax** [comment]
Hi @rgerhards, sorry for the inactivity !
I will try to look into it !

I can already tell you that the change requested has become obsolete as the mmcapture module has been removed from the PR, could you close it ?
I will look into the other issues.

**Comment by rgerhards** [comment]
> I can already tell you that the change requested has become obsolete as the mmcapture module has been removed from the PR, could you close it ?

pls ignore the change request for now - I can't simply remove it, because it would automatically switch to "accept" what I do not want to happen.

Please also have a look at the conflicts CI reports. You need to solve them so that CI can be processed. Thx!

**Comment by rgerhards** [comment]
@frikilax in case you didn't notice, there are quite some CI failures

**Comment by frikilax** [comment]
Yes, I noticed but didn't have much time to look into it, sorry !
I'll try to take some time by the end of the week

**Comment by rgerhards** [comment]
FYI: You need to fix the errors in order to continue on.

**Comment by frikilax** [comment]
I'm conscious this PR has become quite a mess.
I plan on closing this one to create 2 separate ones, with separate topics and correct squashing: one for impcap and another one for ommongodb improvements.

**Comment by rgerhards** [comment]
> I plan on closing this one to create 2 separate ones, with separate topics and correct squashing: one for impcap and another one for ommongodb improvements.

TBH, I think this is a very good plan - makes it much easier to follow things. Also, AFAIK, the mongodb changes are more or less good to merge but are blocked by the others.

**Comment by rgerhards** [comment]
replaced by #3987 and #3988

**Pull request closed without merging**